# MM&A Supermarket Case

## 1. Exploring the Data (EDA)

In [1]:
# importing necessary libraries
import numpy as np
import pandas as pd

In [3]:
# reading in the csv file into a dataframe
file = '../mma_mart.csv'
df = pd.read_csv(file)

In [4]:
# observing the data structure
df.head()

order_id  product_id                                   product_name  \
0         1       49302                               Bulgarian Yogurt   
1         1       11109  Organic 4% Milk Fat Whole Milk Cottage Cheese   
2         1       10246                          Organic Celery Hearts   
3         1       49683                                 Cucumber Kirby   
4         1       43633           Lightly Smoked Sardines in Olive Oil   

   aisle_id                 aisle  department_id    department  
0       120                yogurt             16    dairy eggs  
1       108  other creams cheeses             16    dairy eggs  
2        83      fresh vegetables              4       produce  
3        83      fresh vegetables              4       produce  
4        95   canned meat seafood             15  canned goods

In [5]:
# finding the shape of the data set (row and column)
df.shape

(987259, 7)

The data shows 987259 rows and 7 columns.

In [6]:
# finding amount of unique entries in each column
df.nunique()

order_id         97833
product_id       35070
product_name     35070
aisle_id           134
aisle              134
department_id       21
department          21
dtype: int64

There are ~98000 orders and ~35000 unique products which is across 134 aisles that belong to 21 departments. This shows that selecting 1000-1200 products for the instabasket aisle is selecting only 1000/35000 = ~3% of the total product selection offered at MM&A supermarket.

Next, exploring the aisle and departments to maybe gain some insight on what products should be considered as refrigerant and frozen products.

In [7]:
df['aisle'].unique()

array(['yogurt', 'other creams cheeses', 'fresh vegetables',
       'canned meat seafood', 'fresh fruits', 'packaged cheese', 'eggs',
       'spices seasonings', 'oils vinegars', 'baking ingredients',
       'doughs gelatins bake mixes', 'spreads',
       'packaged vegetables fruits', 'soy lactosefree', 'poultry counter',
       'bread', 'breakfast bakery', 'cold flu allergy',
       'energy granola bars', 'breakfast bars pastries', 'chips pretzels',
       'trail mix snack mix', 'crackers', 'refrigerated',
       'energy sports drinks', 'salad dressing toppings',
       'prepared soups salads', 'milk', 'paper goods',
       'water seltzer sparkling water', 'kosher foods',
       'packaged poultry', 'instant foods', 'packaged produce',
       'cookies cakes', 'candy chocolate', 'body lotions soap',
       'dry pasta', 'laundry', 'air fresheners candles', 'frozen produce',
       'buns rolls', 'canned fruit applesauce', 'juice nectars',
       'granola', 'fresh herbs', 'baby food formul

As can be seen from the aisle names, there are aisles with "frozen" such as `frozen produce` or `frozen meat seafood` in it which all products listed under these aisles would be considered for the 100 frozen product for instabasket. The same can't be said for refridgerant as there a lot of aisles to parse through so it might be better to look at the department to determine what department contains refridgerant products.

In [8]:
df['department'].unique()

array(['dairy eggs', 'produce', 'canned goods', 'pantry', 'meat seafood',
       'bakery', 'personal care', 'snacks', 'breakfast', 'beverages',
       'deli', 'household', 'international', 'dry goods pasta', 'frozen',
       'babies', 'pets', 'alcohol', 'bulk', 'missing', 'other'],
      dtype=object)

From the look at department names, `dairy eggs`, `produce`, `meat seafood`, `deli`, `frozen`, `missing` and `other` are all potential products for fridge and freezer space so it might be better to take a closer look at the products in those department.

In [9]:
dept_list = ['dairy eggs', 'produce', 'meat seafood', 'deli', 'frozen', 'missing', 'other']

In [10]:
# looking at products and aisles of those products for insight as well as how many products each department had
for department in dept_list:
    print(f"Department: {department}")
    print(f"Number of products: {df[df['department']==department]['product_name'].nunique()}\n")
    display(df[df['department']==department][['product_name', 'aisle']])

Department: dairy eggs
Number of products: 2886



product_name  \
0                                        Bulgarian Yogurt   
1           Organic 4% Milk Fat Whole Milk Cottage Cheese   
7                             Organic Whole String Cheese   
8                                      Organic Egg Whites   
17      Total 2% with Strawberry Lowfat Greek Strained...   
...                                                   ...   
987240                        Large Grade AA Organic Eggs   
987241                Reduced Fat Mozarella String Cheese   
987242                   Vanilla Light & Fit Greek Yogurt   
987243               Non-Fat Vanilla Blended Greek Yogurt   
987254            Total 2% All Natural Plain Greek Yogurt   

                       aisle  
0                     yogurt  
1       other creams cheeses  
7            packaged cheese  
8                       eggs  
17                    yogurt  
...                      ...  
987240                  eggs  
987241       packaged cheese  
987242                yogurt  
987243                yogurt  
987254                yogurt  

[164468 rows x 2 columns]

Department: produce
Number of products: 1437



product_name                       aisle
2           Organic Celery Hearts            fresh vegetables
3                  Cucumber Kirby            fresh vegetables
5          Bag of Organic Bananas                fresh fruits
6            Organic Hass Avocado                fresh fruits
9           Michigan Organic Kale            fresh vegetables
...                           ...                         ...
987245                Gala Apples                fresh fruits
987246       Organic Yellow Onion            fresh vegetables
987247       Organic Baby Carrots  packaged vegetables fruits
987249       Organic Baby Spinach  packaged vegetables fruits
987250  Garnet Sweet Potato (Yam)            fresh vegetables

[288775 rows x 2 columns]

Department: meat seafood
Number of products: 692



product_name  \
23      Air Chilled Organic Boneless Skinless Chicken ...   
46               Boneless Skinless Chicken Breast Fillets   
97                            Boneless Beef Sirloin Steak   
176                                      Low Sodium Bacon   
204                  Boneless And Skinless Chicken Breast   
...                                                   ...   
986977                  Organic Air Chilled Whole Chicken   
987067  Natural Hickory Smoked Canadian Bacon Center C...   
987068  All Natural Boneless & Skinless Chicken Breast...   
987175                  Organic Air Chilled Whole Chicken   
987179                      Organic Beef Uncured Hot Dogs   

                         aisle  
23             poultry counter  
46            packaged poultry  
97                meat counter  
176     hot dogs bacon sausage  
204            poultry counter  
...                        ...  
986977         poultry counter  
987067  hot dogs bacon sausage  
987068        packaged poultry  
987175         poultry counter  
987179  hot dogs bacon sausage  

[21588 rows x 2 columns]

Department: deli
Number of products: 1069



product_name                   aisle
40                           Fresh Fruit Salad   prepared soups salads
101                      Mango Pineapple Salsa    fresh dips tapenades
179     Basil, Asiago & Pine Nut Pesto Ravioli          prepared meals
191                             Yuba Tofu Skin  tofu meat alternatives
192                          Organic Firm Tofu  tofu meat alternatives
...                                        ...                     ...
987153                    Spicy Avocado Hummus    fresh dips tapenades
987158                               Guacamole    fresh dips tapenades
987199                           Guacamole Dip    fresh dips tapenades
987204                         Original Hummus    fresh dips tapenades
987253                         Original Hummus    fresh dips tapenades

[32008 rows x 2 columns]

Department: frozen
Number of products: 3127



product_name                    aisle
68                                   Pineapple Chunks           frozen produce
100            Teriyaki & Pineapple Chicken Meatballs             frozen meals
114     All Natural Boneless Skinless Chicken Breasts      frozen meat seafood
118                           Combination Pizza Rolls  frozen appetizers sides
135                    Organic Mini Homestyle Waffles         frozen breakfast
...                                               ...                      ...
987176                 Organic Ice Cream Vanilla Bean            ice cream ice
987180   Dairy Free Coconut Milk Frozen Dessert Minis            ice cream ice
987184                           Organic Mango Chunks           frozen produce
987210                            Bag of Large Lemons      frozen meat seafood
987215                              Vanilla Ice Cream            ice cream ice

[68202 rows x 2 columns]

Department: missing
Number of products: 518



product_name    aisle
654                          Tomato Basil Bisque Soup  missing
1511      Cold Pressed Watermelon & Lemon Juice Blend  missing
1512                           Paleo Blueberry Muffin  missing
2126    Magic Tape Refillable Dispenser 3/4\" x 850\"  missing
4121                           Organic Poblano Pepper  missing
...                                               ...      ...
986067                       Oatneal Cookie Ice Cream  missing
986086    Dairy Free Unsweetened Almond Milk Beverage  missing
986103                Organic Asian Chopped Salad Kit  missing
986275                                      Lemon Bag  missing
986276      Organic Nondairy Strawberry Cashew Yogurt  missing

[2289 rows x 2 columns]

Department: other
Number of products: 303



product_name  aisle
691            Coffee Mate French Vanilla Creamer Packets  other
1077                        SleepGels Nighttime Sleep Aid  other
1926                             Roasted Unsalted Almonds  other
1985                         Camilia, Single Liquid Doses  other
2127    Maximum Strength Original Paste Diaper Rash Oi...  other
...                                                   ...    ...
983992                           Roasted Unsalted Almonds  other
984112                           Boneless Pork Tenderloin  other
984515                              Roasted Almond Butter  other
984534            Light CocoWhip! Coconut Whipped Topping  other
986326                                   93/7 Ground Beef  other

[1097 rows x 2 columns]

Looking at the results, products from `dairy eggs`, `meat seafood` `deli`, and some products from `missing` and `other` (would need to look at what exactly the product is since there was "Oatneal Cookie Ice Cream" in `missing` aisle which is a frozen item and "Boneless Pork Tenderloin" in `other` which is a `meat seafood` item) would be the departments that require refridgerant. Produce might not be a neccisity to be refridgerated as it can be stored at room temperature but only refridgerated to prolong shelf life so if there is a lot of demand for the produce, then it wouldn't be on the shelf for long but if there is low demand for it, then it might be better to not even have it as a selection so it might be a better idea to give the refridgerator space to the other aisles.

In [11]:
# Sanity check to see if aisles such as frozen produce are part of the frozen department or produce department
frozen_aisles = df[df['aisle'].str.contains('frozen', case=False)]['aisle'].unique()
print(frozen_aisles)

['frozen produce' 'frozen meals' 'frozen meat seafood'
 'frozen appetizers sides' 'frozen breakfast' 'frozen breads doughs'
 'frozen vegan vegetarian' 'frozen pizza' 'frozen dessert' 'frozen juice']


In [12]:
frozen_department = df[df['aisle'].str.contains('frozen', case=False)]['department'].unique()
print(frozen_department)

['frozen']


It is safe to assume that 100 products for the frozen products will be from the frozen department and maybe `missing` and `other` department.

In [26]:
# storing all frozen products into a variable
freezer_df = df[df['department']=='frozen']
freezer_df

order_id  product_id                                   product_name  \
68             7       46802                               Pineapple Chunks   
100           11       30162         Teriyaki & Pineapple Chicken Meatballs   
114           12       38050  All Natural Boneless Skinless Chicken Breasts   
118           12       29471                        Combination Pizza Rolls   
135           14         162                 Organic Mini Homestyle Waffles   
...          ...         ...                                            ...   
987176     99993       43749                 Organic Ice Cream Vanilla Bean   
987180     99993        5537   Dairy Free Coconut Milk Frozen Dessert Minis   
987184     99993       26128                           Organic Mango Chunks   
987210     99996        9434                            Bag of Large Lemons   
987215     99996       32691                              Vanilla Ice Cream   

        aisle_id                    aisle  department_id department  
68           116           frozen produce              1     frozen  
100           38             frozen meals              1     frozen  
114           34      frozen meat seafood              1     frozen  
118          129  frozen appetizers sides              1     frozen  
135           52         frozen breakfast              1     frozen  
...          ...                      ...            ...        ...  
987176        37            ice cream ice              1     frozen  
987180        37            ice cream ice              1     frozen  
987184       116           frozen produce              1     frozen  
987210        34      frozen meat seafood              1     frozen  
987215        37            ice cream ice              1     frozen  

[68202 rows x 7 columns]

In [14]:
# storing all fridge products into a variable
fridge_df = df[df['department'].isin(['dairy eggs', 'meat seafood', 'deli'])]
fridge_df

order_id  product_id  \
0              1       49302   
1              1       11109   
7              1       22035   
8              2       33120   
17             3       33754   
...          ...         ...   
987241     99999       10034   
987242     99999       44085   
987243     99999        9558   
987253    100000       30489   
987254    100000       30169   

                                             product_name  aisle_id  \
0                                        Bulgarian Yogurt       120   
1           Organic 4% Milk Fat Whole Milk Cottage Cheese       108   
7                             Organic Whole String Cheese        21   
8                                      Organic Egg Whites        86   
17      Total 2% with Strawberry Lowfat Greek Strained...       120   
...                                                   ...       ...   
987241                Reduced Fat Mozarella String Cheese        21   
987242                   Vanilla Light & Fit Greek Yogurt       120   
987243               Non-Fat Vanilla Blended Greek Yogurt       120   
987253                                    Original Hummus        67   
987254            Total 2% All Natural Plain Greek Yogurt       120   

                       aisle  department_id  department  
0                     yogurt             16  dairy eggs  
1       other creams cheeses             16  dairy eggs  
7            packaged cheese             16  dairy eggs  
8                       eggs             16  dairy eggs  
17                    yogurt             16  dairy eggs  
...                      ...            ...         ...  
987241       packaged cheese             16  dairy eggs  
987242                yogurt             16  dairy eggs  
987243                yogurt             16  dairy eggs  
987253  fresh dips tapenades             20        deli  
987254                yogurt             16  dairy eggs  

[218064 rows x 7 columns]

In [16]:
# removing the freezer and fridge products from the main df as this will be for the 1000 in aisle product selection
remove_index = freezer_df.index.union(fridge_df.index)
filtered_df = df.drop(remove_index)
filtered_df

order_id  product_id                            product_name  \
2              1       10246                   Organic Celery Hearts   
3              1       49683                          Cucumber Kirby   
4              1       43633    Lightly Smoked Sardines in Olive Oil   
5              1       13176                  Bag of Organic Bananas   
6              1       47209                    Organic Hass Avocado   
...          ...         ...                                     ...   
987252    100000       19508                          Corn Tortillas   
987255    100000       38734                    Wheat Sandwich Thins   
987256    100000       36759  Unscented Long Lasting Stick Deodorant   
987257    100000       37107                            Ground Cumin   
987258    100000       31506                  Extra Virgin Olive Oil   

        aisle_id                 aisle  department_id     department  
2             83      fresh vegetables              4        produce  
3             83      fresh vegetables              4        produce  
4             95   canned meat seafood             15   canned goods  
5             24          fresh fruits              4        produce  
6             24          fresh fruits              4        produce  
...          ...                   ...            ...            ...  
987252       128  tortillas flat bread              3         bakery  
987255       128  tortillas flat bread              3         bakery  
987256        80            deodorants             11  personal care  
987257       104     spices seasonings             13         pantry  
987258        19         oils vinegars             13         pantry  

[700993 rows x 7 columns]

In [18]:
# Sanity Check to see if the rows were removed correctly
freezer_df.shape[0] + fridge_df.shape[0] + filtered_df.shape[0] == df.shape[0]

True

In [19]:
# Check to see if the right departments are removed
filtered_df['department'].unique()

array(['produce', 'canned goods', 'pantry', 'bakery', 'personal care',
       'snacks', 'breakfast', 'beverages', 'household', 'international',
       'dry goods pasta', 'babies', 'pets', 'alcohol', 'bulk', 'missing',
       'other'], dtype=object)

It did remove the aisle 

## 2. Product Selection

In [20]:
# import the library needed to split
from sklearn.model_selection import train_test_split

# splitting the data 
train_data, test_data = train_test_split(df, test_size=0.2, random_state=1)

In [22]:
# Sanity check
train_data.head()

order_id  product_id  \
893885     90781       33731   
51142       5194        7051   
738442     74808       47626   
6338         641       47626   
922685     93610        8424   

                                             product_name  aisle_id  \
893885                                    Grated Parmesan        21   
51142   Happy Tot Banana, Peach, Prune & Coconut Organ...        92   
738442                                        Large Lemon        24   
6338                                          Large Lemon        24   
922685                                     Broccoli Crown        83   

                    aisle  department_id  department  
893885    packaged cheese             16  dairy eggs  
51142   baby food formula             18      babies  
738442       fresh fruits              4     produce  
6338         fresh fruits              4     produce  
922685   fresh vegetables              4     produce

In [27]:
# Sanity Check
test_data.head()

order_id  product_id  \
138856     14036       10321   
918287     93161       46584   
154109     15573        2228   
955032     96760       23695   
226812     22918        7781   

                                             product_name  aisle_id  \
138856           Grands! Flaky Layers Buttermilk Biscuits       105   
918287       YoKids Blueberry & Strawberry/Vanilla Yogurt       120   
154109                        Organic Frozen Mango Chunks       116   
955032                           California Veggie Burger        42   
226812  Organic Sticks Low Moisture Part Skim Mozzarel...        21   

                             aisle  department_id  department  
138856  doughs gelatins bake mixes             13      pantry  
918287                      yogurt             16  dairy eggs  
154109              frozen produce              1      frozen  
955032     frozen vegan vegetarian              1      frozen  
226812             packaged cheese             16  dairy eggs

In [70]:
# Checking to see if a product appears multiple time in one order if multiple of the product is ordered 
df[df[['order_id', 'product_id']].duplicated()]

Empty DataFrame
Columns: [order_id, product_id, product_name, aisle_id, aisle, department_id, department]
Index: []

### Method 2: Product Selection Proportional to Aisle Selection

This second method is to get a variety of items from each aisle proportional to the amount of orders for that aisle.

In [71]:
# finding the total amount of products ordered
train_size = train_data.shape[0]
train_size

789807

In [72]:
# finding the amount of items ordered for each aisle
aisle_df=train_data.groupby('aisle').size().reset_index(name='count')\
    .sort_values(by='count', ascending=False)
aisle_df

aisle  count
50                 fresh fruits  88835
53             fresh vegetables  83118
98   packaged vegetables fruits  43197
133                      yogurt  35165
93              packaged cheese  23726
..                          ...    ...
76             kitchen supplies    213
2              baby accessories    186
3           baby bath body care    184
8                        beauty    134
58                 frozen juice    116

[134 rows x 2 columns]

In [73]:
aisle_df['percentage']=aisle_df['count']/train_size*100
aisle_df

aisle  count  percentage
50                 fresh fruits  88835   11.247685
53             fresh vegetables  83118   10.523837
98   packaged vegetables fruits  43197    5.469311
133                      yogurt  35165    4.452354
93              packaged cheese  23726    3.004025
..                          ...    ...         ...
76             kitchen supplies    213    0.026969
2              baby accessories    186    0.023550
3           baby bath body care    184    0.023297
8                        beauty    134    0.016966
58                 frozen juice    116    0.014687

[134 rows x 3 columns]

In [74]:
aisle_df

aisle  count  percentage
50                 fresh fruits  88835   11.247685
53             fresh vegetables  83118   10.523837
98   packaged vegetables fruits  43197    5.469311
133                      yogurt  35165    4.452354
93              packaged cheese  23726    3.004025
..                          ...    ...         ...
76             kitchen supplies    213    0.026969
2              baby accessories    186    0.023550
3           baby bath body care    184    0.023297
8                        beauty    134    0.016966
58                 frozen juice    116    0.014687

[134 rows x 3 columns]

In [75]:
train_aisle_mapping = train_data[['aisle', 'department']]
train_aisle_mapping = train_aisle_mapping.drop_duplicates()
train_aisle_mapping

aisle     department
893885               packaged cheese     dairy eggs
51142              baby food formula         babies
738442                  fresh fruits        produce
922685              fresh vegetables        produce
745108                          milk     dairy eggs
...                              ...            ...
395417            ice cream toppings         snacks
792140           baby bath body care         babies
740798              kitchen supplies      household
205099  bulk grains rice dried goods           bulk
905656                     first aid  personal care

[134 rows x 2 columns]

In [76]:
# Merge the 'aisle_count' DataFrame with the 'department' column from 'train_data' based on the 'aisle' column
merged_df = aisle_df.merge(train_aisle_mapping, on='aisle', how='left')
merged_df

aisle  count  percentage     department
0                  fresh fruits  88835   11.247685        produce
1              fresh vegetables  83118   10.523837        produce
2    packaged vegetables fruits  43197    5.469311        produce
3                        yogurt  35165    4.452354     dairy eggs
4               packaged cheese  23726    3.004025     dairy eggs
..                          ...    ...         ...            ...
129            kitchen supplies    213    0.026969      household
130            baby accessories    186    0.023550         babies
131         baby bath body care    184    0.023297         babies
132                      beauty    134    0.016966  personal care
133                frozen juice    116    0.014687         frozen

[134 rows x 4 columns]

In [77]:
merged_df['department'].unique()

array(['produce', 'dairy eggs', 'beverages', 'snacks', 'bakery', 'frozen',
       'deli', 'babies', 'breakfast', 'canned goods', 'pantry',
       'meat seafood', 'dry goods pasta', 'household', 'international',
       'missing', 'personal care', 'pets', 'alcohol', 'other', 'bulk'],
      dtype=object)

As can be seen aisle 

In [62]:
train_data.head()

order_id  product_id  \
893885     90781       33731   
51142       5194        7051   
738442     74808       47626   
6338         641       47626   
922685     93610        8424   

                                             product_name  aisle_id  \
893885                                    Grated Parmesan        21   
51142   Happy Tot Banana, Peach, Prune & Coconut Organ...        92   
738442                                        Large Lemon        24   
6338                                          Large Lemon        24   
922685                                     Broccoli Crown        83   

                    aisle  department_id  department  
893885    packaged cheese             16  dairy eggs  
51142   baby food formula             18      babies  
738442       fresh fruits              4     produce  
6338         fresh fruits              4     produce  
922685   fresh vegetables              4     produce

In [78]:
# storing all frozen products into a variable
freezer_df = train_data[train_data['department']=='frozen']
freezer_df


order_id  product_id                               product_name  \
86782       8859       30675                   Edamame Soybeans in Pods   
584410     59226       44753      New York Super Fudge Chunk® Ice Cream   
447570     45412        1609  Egg, Uncured Bacon & Cheese Scramble Cups   
622031     63014       13378                        Stir Fry Vegetables   
244156     24680        5449                           Margherita Pizza   
...          ...         ...                                        ...   
128896     13044        1986    Steamers Healthy Vision Vegetable Blend   
723457     73315       38274               Ice Cream Sandwiches Vanilla   
699870     70944       38959               Smoothies, Strawberries Wild   
21758       2231        2164            Waffles, Sweet Belgian, Vanilla   
436973     44314        9595             Organic Homestyle Mini Waffles   

        aisle_id             aisle  department_id department  
86782        116    frozen produce              1     frozen  
584410        37     ice cream ice              1     frozen  
447570        38      frozen meals              1     frozen  
622031       116    frozen produce              1     frozen  
244156        79      frozen pizza              1     frozen  
...          ...               ...            ...        ...  
128896       116    frozen produce              1     frozen  
723457        37     ice cream ice              1     frozen  
699870       113      frozen juice              1     frozen  
21758         52  frozen breakfast              1     frozen  
436973        52  frozen breakfast              1     frozen  

[54567 rows x 7 columns]

In [84]:
freezer_df.shape[0]/train_data.shape[0]*100

6.9089030611275914

6% only are being sold at frozen department

In [108]:
freezer_df[freezer_df['product_name'] == 'Blueberries']

order_id  product_id product_name  aisle_id           aisle  \
736916     74650        9076  Blueberries       116  frozen produce   
962996     97578        9076  Blueberries       116  frozen produce   
930393     94350        9076  Blueberries       116  frozen produce   
428923     43481        9076  Blueberries       116  frozen produce   
532596     53938        9076  Blueberries       116  frozen produce   
...          ...         ...          ...       ...             ...   
566057     57373        9076  Blueberries       116  frozen produce   
263096     26536        9076  Blueberries       116  frozen produce   
645086     65381        9076  Blueberries       116  frozen produce   
348165     35220        9076  Blueberries       116  frozen produce   
99782      10121        9076  Blueberries       116  frozen produce   

        department_id department  
736916              1     frozen  
962996              1     frozen  
930393              1     frozen  
428923              1     frozen  
532596              1     frozen  
...               ...        ...  
566057              1     frozen  
263096              1     frozen  
645086              1     frozen  
348165              1     frozen  
99782               1     frozen  

[1404 rows x 7 columns]

In [89]:
# let us see the most selling product in frozen department
freezer_df['product_name'].value_counts()

Blueberries                                   1404
Organic Broccoli Florets                       845
Organic Whole Strawberries                     792
Frozen Organic Wild Blueberries                740
Pineapple Chunks                               568
                                              ... 
Regular Whipped Topping                          1
Restaurant Quality & Oven Ready Crab Cakes       1
Lite Whipped Topping                             1
Beer Battered Golden Whole Fish Fillets          1
Lemon Ice Cubes                                  1
Name: product_name, Length: 3013, dtype: int64

In [132]:
# Shows the highest 100 selling product
top_freezer = freezer_df.groupby(['aisle', 'product_name', 'department']).size().reset_index(name='count')\
    .sort_values(by='count', ascending=False).head(100)
top_freezer

aisle                              product_name  \
1763           frozen produce                               Blueberries   
1893           frozen produce                  Organic Broccoli Florets   
1943           frozen produce                Organic Whole Strawberries   
1848           frozen produce           Frozen Organic Wild Blueberries   
1962           frozen produce                          Pineapple Chunks   
...                       ...                                       ...   
2084  frozen vegan vegetarian                            Chik'n Nuggets   
1051             frozen meals              Light & Lean Spinach Lasagna   
2951            ice cream ice                    Vanilla Bean Ice Cream   
625            frozen dessert             Dark Chocolate Covered Banana   
2345            ice cream ice  Coconut Almond Minis Frozen Dessert Bars   

     department  count  
1763     frozen   1404  
1893     frozen    845  
1943     frozen    792  
1848     frozen    740  
1962     frozen    568  
...         ...    ...  
2084     frozen    102  
1051     frozen    101  
2951     frozen    101  
625      frozen    100  
2345     frozen     99  

[100 rows x 4 columns]

In [129]:
# Let's see how many product are from one aisle
freezer_value_count = top_freezer['aisle'].value_counts()
pd.DataFrame({'aisle': freezer_value_count.index, 'count': freezer_value_count.values})

aisle  count
0           frozen produce     26
1            ice cream ice     22
2         frozen breakfast     13
3             frozen meals     13
4  frozen appetizers sides     10
5             frozen pizza      7
6      frozen meat seafood      4
7  frozen vegan vegetarian      3
8     frozen breads doughs      1
9           frozen dessert      1

The insight shows that `frozen produce` and `ice cream ice` are the highest selling point.

In [79]:
# storing all fridge products into a variable
fridge_df = train_data[train_data['department'].isin(['dairy eggs', 'meat seafood', 'deli'])]
fridge_df

order_id  product_id  \
893885     90781       33731   
745108     75500       43693   
713260     72291       33791   
316928     32017        2344   
856602     86898       29118   
...          ...         ...   
167302     16901       37388   
229520     23210        7781   
371403     37543       27845   
491263     49775       42002   
470924     47761       42828   

                                             product_name  aisle_id  \
893885                                    Grated Parmesan        21   
745108                               Whole Vitamin D Milk        84   
713260        Raspberry on the Bottom Nonfat Greek Yogurt       120   
316928                Organic SprouTofu Savory Baked Tofu        14   
856602                         French Vanilla Soy Creamer        91   
...                                                   ...       ...   
167302                         Cream Top Blueberry Yogurt       120   
229520  Organic Sticks Low Moisture Part Skim Mozzarel...        21   
371403                                 Organic Whole Milk        84   
491263                       Pasture Raised Salted Butter        36   
470924                               Whipped Cream Cheese       108   

                         aisle  department_id  department  
893885         packaged cheese             16  dairy eggs  
745108                    milk             16  dairy eggs  
713260                  yogurt             16  dairy eggs  
316928  tofu meat alternatives             20        deli  
856602         soy lactosefree             16  dairy eggs  
...                        ...            ...         ...  
167302                  yogurt             16  dairy eggs  
229520         packaged cheese             16  dairy eggs  
371403                    milk             16  dairy eggs  
491263                  butter             16  dairy eggs  
470924    other creams cheeses             16  dairy eggs  

[174379 rows x 7 columns]

In [ ]:
fridge_df.shape[0]/train_data.shape[0]*100

22.078685045840313

22% are being sold at fridge department which is more than frozen food department

In [117]:
# let us see the most selling product in frozen department
fridge_df['product_name'].value_counts()

Organic Whole Milk                                   3306
Organic Half & Half                                  1920
Original Hummus                                      1702
Half & Half                                          1636
Organic Whole String Cheese                          1455
                                                     ... 
Original Blackberry Harvest Low Fat Yogurt              1
Chicken Kiev                                            1
Creamy Chocolate Sugar Free Powder Coffee Creamer       1
Non Alcoholic French Vanilla Coffee Creamer             1
Boneless Beef Short Ribs                                1
Name: product_name, Length: 4480, dtype: int64

In [131]:
# Shows the highest 100 selling product
top_fridge = fridge_df.groupby(['aisle', 'product_name', 'department']).size().reset_index(name='count')\
    .sort_values(by='count', ascending=False).head(100)

top_fridge

aisle                             product_name  \
1435                    milk                       Organic Whole Milk   
196                    cream                      Organic Half & Half   
532     fresh dips tapenades                          Original Hummus   
163                    cream                              Half & Half   
2127         packaged cheese              Organic Whole String Cheese   
...                      ...                                      ...   
801   hot dogs bacon sausage            Organic Beef Uncured Hot Dogs   
4155                  yogurt          Organic Plain Whole Milk Yogurt   
1209              lunch meat                  Uncured Slow Cooked Ham   
318                     eggs  Organic Extra Large Grade AA Brown Eggs   
810   hot dogs bacon sausage                     Organic Sunday Bacon   

        department  count  
1435    dairy eggs   3306  
196     dairy eggs   1920  
532           deli   1702  
163     dairy eggs   1636  
2127    dairy eggs   1455  
...            ...    ...  
801   meat seafood    325  
4155    dairy eggs    324  
1209          deli    323  
318     dairy eggs    320  
810   meat seafood    316  

[100 rows x 4 columns]

In [133]:
# Let's see how many product are from one aisle
fridge_value_count = top_fridge['aisle'].value_counts()
pd.DataFrame({'aisle': fridge_value_count.index, 'count': fridge_value_count.values})

aisle  count
0                   yogurt     19
1                     milk     16
2          packaged cheese     13
3                     eggs      9
4          soy lactosefree      8
5                   butter      6
6   hot dogs bacon sausage      6
7     fresh dips tapenades      5
8               lunch meat      5
9                    cream      4
10    other creams cheeses      3
11         poultry counter      3
12        packaged poultry      1
13  tofu meat alternatives      1
14         seafood counter      1

The insight shows that `yogurt`, `packaged cheese` and `milk` are the highest selling point.

In [136]:
combine_frozen_fridge = pd.concat([top_freezer,top_fridge], ignore_index=True)

In [137]:
combine_frozen_fridge

aisle                             product_name  \
0            frozen produce                              Blueberries   
1            frozen produce                 Organic Broccoli Florets   
2            frozen produce               Organic Whole Strawberries   
3            frozen produce          Frozen Organic Wild Blueberries   
4            frozen produce                         Pineapple Chunks   
..                      ...                                      ...   
195  hot dogs bacon sausage            Organic Beef Uncured Hot Dogs   
196                  yogurt          Organic Plain Whole Milk Yogurt   
197              lunch meat                  Uncured Slow Cooked Ham   
198                    eggs  Organic Extra Large Grade AA Brown Eggs   
199  hot dogs bacon sausage                     Organic Sunday Bacon   

       department  count  
0          frozen   1404  
1          frozen    845  
2          frozen    792  
3          frozen    740  
4          frozen    568  
..            ...    ...  
195  meat seafood    325  
196    dairy eggs    324  
197          deli    323  
198    dairy eggs    320  
199  meat seafood    316  

[200 rows x 4 columns]

We combined 200 products in demand for 2 departments `frozen` and `refrigerator`

In [138]:
# Combined products which shows number of product in each aisle
value_count = combine_frozen_fridge['aisle'].value_counts()
pd.DataFrame({'aisle': value_count.index, 'count': value_count.values})

aisle  count
0            frozen produce     26
1             ice cream ice     22
2                    yogurt     19
3                      milk     16
4              frozen meals     13
5          frozen breakfast     13
6           packaged cheese     13
7   frozen appetizers sides     10
8                      eggs      9
9           soy lactosefree      8
10             frozen pizza      7
11   hot dogs bacon sausage      6
12                   butter      6
13               lunch meat      5
14     fresh dips tapenades      5
15                    cream      4
16      frozen meat seafood      4
17  frozen vegan vegetarian      3
18     other creams cheeses      3
19          poultry counter      3
20         packaged poultry      1
21           frozen dessert      1
22   tofu meat alternatives      1
23     frozen breads doughs      1
24          seafood counter      1

In [146]:
other_df = train_data[~train_data['department'].isin(['dairy eggs', 'meat seafood', 'deli', 'frozen'])]
other_df

order_id  product_id  \
51142       5194        7051   
738442     74808       47626   
6338         641       47626   
922685     93610        8424   
533757     54051       23165   
...          ...         ...   
73349       7471       18963   
836489     84828       16665   
791624     80274       18628   
491755     49823       48946   
128037     12963       38256   

                                             product_name  aisle_id  \
51142   Happy Tot Banana, Peach, Prune & Coconut Organ...        92   
738442                                        Large Lemon        24   
6338                                          Large Lemon        24   
922685                                     Broccoli Crown        83   
533757                                       Organic Leek        83   
...                                                   ...       ...   
73349                                 Apple Cider Vinegar        19   
836489                          Banana Bread Sprouted Bar         3   
791624             Organic Volcano Rice Ready To Eat Bowl         4   
491755                                      Earl Grey Tea        94   
128037                  Maple Glazed Pecan & Sea Salt Bar         3   

                      aisle  department_id       department  
51142     baby food formula             18           babies  
738442         fresh fruits              4          produce  
6338           fresh fruits              4          produce  
922685     fresh vegetables              4          produce  
533757     fresh vegetables              4          produce  
...                     ...            ...              ...  
73349         oils vinegars             13           pantry  
836489  energy granola bars             19           snacks  
791624        instant foods              9  dry goods pasta  
491755                  tea              7        beverages  
128037  energy granola bars             19           snacks  

[560861 rows x 7 columns]

In [150]:
other_df.shape[0]/train_data.shape[0]*100

71.0124118930321

71% are being sold at fridge department which is more than frozen food department

In [151]:
# let us see the most selling product in frozen department
other_df['product_name'].value_counts()

Banana                                                        11640
Bag of Organic Bananas                                         9396
Organic Strawberries                                           6484
Organic Baby Spinach                                           5849
Organic Hass Avocado                                           5102
                                                              ...  
Original Spiced Rum                                               1
Chocolate Peanut Butter Powder                                    1
Multi-Faceted Shimmering Colour 21 Bright Black Hair Color        1
Chicken Flavor Chow Mein Noodles                                  1
All Natural Table Syrup                                           1
Name: product_name, Length: 25669, dtype: int64

In [153]:
# Shows the highest 100 selling product
top_other = other_df.groupby(['aisle', 'product_name', 'department']).size().reset_index(name='count')\
    .sort_values(by='count', ascending=False).head(800)

top_other

aisle                                product_name  \
12288                fresh fruits                                      Banana   
12286                fresh fruits                      Bag of Organic Bananas   
12495                fresh fruits                        Organic Strawberries   
18856  packaged vegetables fruits                        Organic Baby Spinach   
12464                fresh fruits                        Organic Hass Avocado   
...                           ...                                         ...   
5124          canned meat seafood          Solid White Albacore Tuna In Water   
3069             breakfast bakery               Organic White English Muffins   
1091            baby food formula  Peach,  Apricot & Banana Stage 2 Baby Food   
1646           baking ingredients                             Organic Pumpkin   
12793            fresh vegetables                      California Cauliflower   

         department  count  
12288       produce  11640  
12286       produce   9396  
12495       produce   6484  
18856       produce   5849  
12464       produce   5102  
...             ...    ...  
5124   canned goods    109  
3069         bakery    109  
1091         babies    108  
1646         pantry    108  
12793       produce    108  

[800 rows x 4 columns]

In [154]:
# Let's see how many product are from one aisle
value_count = top_other['aisle'].value_counts()
pd.DataFrame({'aisle': value_count.index, 'count': value_count.values})

aisle  count
0                fresh vegetables    134
1                    fresh fruits     79
2      packaged vegetables fruits     53
3   water seltzer sparkling water     39
4                  chips pretzels     35
..                            ...    ...
60                        missing      1
61     marinades meat preparation      1
62   bulk dried fruits vegetables      1
63                   latino foods      1
64            canned meat seafood      1

[65 rows x 2 columns]

In [157]:
combine_df = pd.concat([top_other,combine_frozen_fridge], ignore_index=True)
combine_df

aisle                             product_name  \
0                  fresh fruits                                   Banana   
1                  fresh fruits                   Bag of Organic Bananas   
2                  fresh fruits                     Organic Strawberries   
3    packaged vegetables fruits                     Organic Baby Spinach   
4                  fresh fruits                     Organic Hass Avocado   
..                          ...                                      ...   
995      hot dogs bacon sausage            Organic Beef Uncured Hot Dogs   
996                      yogurt          Organic Plain Whole Milk Yogurt   
997                  lunch meat                  Uncured Slow Cooked Ham   
998                        eggs  Organic Extra Large Grade AA Brown Eggs   
999      hot dogs bacon sausage                     Organic Sunday Bacon   

       department  count  
0         produce  11640  
1         produce   9396  
2         produce   6484  
3         produce   5849  
4         produce   5102  
..            ...    ...  
995  meat seafood    325  
996    dairy eggs    324  
997          deli    323  
998    dairy eggs    320  
999  meat seafood    316  

[1000 rows x 4 columns]

In [158]:
# Let's see how many product are from one aisle
value_count = combine_df['aisle'].value_counts()
pd.DataFrame({'aisle': value_count.index, 'count': value_count.values})

aisle  count
0                fresh vegetables    134
1                    fresh fruits     79
2      packaged vegetables fruits     53
3   water seltzer sparkling water     39
4                  chips pretzels     35
..                            ...    ...
85                        spirits      1
86                        laundry      1
87                  beers coolers      1
88           frozen breads doughs      1
89                seafood counter      1

[90 rows x 2 columns]

In [161]:
# Let's see how many product are from one aisle
department_value_count = combine_df['department'].value_counts()
pd.DataFrame({'department': department_value_count.index, 'count': department_value_count.values})

department  count
0           produce    297
1            snacks    122
2         beverages    114
3            frozen    100
4        dairy eggs     78
5            pantry     62
6            bakery     59
7      canned goods     42
8   dry goods pasta     39
9         breakfast     27
10           babies     13
11        household     12
12             deli     11
13     meat seafood     11
14          alcohol      5
15    personal care      2
16    international      2
17             bulk      2
18            other      1
19          missing      1

In [163]:
# count frozen products
combine_df[combine_df['department'] == 'frozen'].count()

aisle           100
product_name    100
department      100
count           100
dtype: int64